### ISSUE TITLE
Rendering bugs and crashes with Raspberry Pi DRI3

### ISSUE THREAD

**Issue by graysky2** [TigerVNC/tigervnc#1790]
**Describe the bug**
Chromium runs but is unusable due to numerous rendering errors under TigerVNC 1.14.0.  Downgrading to TigerVNC 1.13.1 fixes the issue.

**To Reproduce**
1. On RPi4B, load chromium in a vncsession

**Expected behavior**
Chromium GUI should be readable and responsive

**Screenshots**
Attached is a screenshot of chromium under 1.14.0 exhibiting the bug:
![chromium](https://github.com/user-attachments/assets/4562dc04-8778-4bbf-8da9-29b466fb968a)

**Client (please complete the following information):**
 - OS: Arch Linux
 - VNC client: TigerVNC
 - VNC client version: 4.14.0
 - Client downloaded from: Official Arch Linux package

**Server (please complete the following information):**
 - OS: Arch ARM
 - VNC server: TigerVNC
 - VNC server version: 4.14.0
 - Server downloaded from: Official Arch ARM package
 - Server was started using: systemd `vncserver@5901.service` which is provided below for reference.

**Additional context**

Here is the output from running `chromium` on the terminal:
```
Xlib:  extension "XFree86-VidModeExtension" missing on display ":1".
[4480:4480:0728/145804.769912:ERROR:gl_display.cc(497)] EGL Driver message (Error) eglGetMscRateANGLE: glXGetMscRateOML failed.
Xlib:  extension "XFree86-VidModeExtension" missing on display ":1".
[4480:4480:0728/145804.951000:ERROR:gl_display.cc(497)] EGL Driver message (Error) eglGetMscRateANGLE: glXGetMscRateOML failed.
Xlib:  extension "XFree86-VidModeExtension" missing on display ":1".
[4480:4480:0728/145805.037749:ERROR:gl_display.cc(497)] EGL Driver message (Error) eglGetMscRateANGLE: glXGetMscRateOML failed.
Xlib:  extension "XFree86-VidModeExtension" missing on display ":1".
[4480:4480:0728/145805.087044:ERROR:gl_display.cc(497)] EGL Driver message (Error) eglGetMscRateANGLE: glXGetMscRateOML failed.
Xlib:  extension "XFree86-VidModeExtension" missing on display ":1".
[4480:4480:0728/145805.155235:ERROR:gl_display.cc(497)] EGL Driver message (Error) eglGetMscRateANGLE: glXGetMscRateOML failed.
[4480:4480:0728/145808.028138:ERROR:gl_display.cc(497)] EGL Driver message (Error) eglGetMscRateANGLE: glXGetMscRateOML failed.
[4480:4480:0728/145808.028846:ERROR:sync_control_vsync_provider.cc(140)] Calculated bogus refresh interval=0.250806 s, last_timebase_=79485183829 bogo-microseconds, timebase=79485685441 bogo-microseconds, last_media_stream_counter_=66, media_stream_counter=68
Xlib:  extension "XFree86-VidModeExtension" missing on display ":1".
```

The systemd user service file `~/.config/systemd/user/vncserver@.service`:
```
[Unit]
Description=Remote desktop service (VNC)
After=syslog.target network.target

[Service]
Type=forking
ExecStartPre=/bin/sh -c '/usr/bin/vncserver -kill %i > /dev/null 2>&1 || :'
ExecStart=/usr/bin/vncserver -geometry 2000x1200 -localhost -alwaysshared %i
ExecStop=/usr/bin/vncserver -kill %i

[Install]
WantedBy=default.target
```

**Comment by oddfellow** [comment]
Some info to perhaps help narrow it down:

tigervnc 1.14.0-1 x86_64 on Arch Linux (client and server)

Although I get the same error message when running chromium from xterm, I do **not** have any visual rendering bugs, everything looks and works normally in chromium.

Some differences to your setup:

- I use the package provided systemd service file to start vncserver (/usr/lib/systemd/system/vncserver@.service)
- x86_64 Arch Linux on the server (and client) side
- i3 as X window manager on the server side in the vnc session
- sway (Wayland) on the client side

**Comment by CendioOssman** [comment]
Maybe some bad interaction with the RPi4 GPU. We unfortunately did not test the new DRI3 stuff on that.

To check if it's that, could you add the `-rendernode ""` argument to the server?

**Comment by graysky2** [comment]
Thanks for the reply.  I modified my service like this:
```
- ExecStart=/usr/bin/vncserver -geometry 2000x1200 -localhost -alwaysshared %i
+ ExecStart=/usr/bin/vncserver -rendernode "" -geometry 2000x1200 -localhost -alwaysshared %i
```

I am not seeing anything in `~/.config/hostname:1.log`:
```
~/.vnc is deprecated, please consult 'man vncsession' for paths to migrate to. at /usr/bin/vncserver line 43.
Using desktop session lxqt

New 'familyroom:1 (squishy)' desktop is familyroom:1

Starting desktop session lxqt


Xvnc TigerVNC 1.14.0 - built Jul 26 2024 23:15:30
Copyright (C) 1999-2024 TigerVNC Team and many others (see README.rst)
See https://www.tigervnc.org for information on TigerVNC.
Underlying X server release 12101013


Mon Jul 29 15:29:05 2024
 vncext:      VNC extension running!
 vncext:      Listening for VNC connections on all interface(s), port 5901
 vncext:      created VNC server for screen 0
[mi] mieq: warning: overriding existing handler (nil) with 0xaaaae973b188 for event 2
[mi] mieq: warning: overriding existing handler (nil) with 0xaaaae973b188 for event 3
xinit: XFree86_VT property unexpectedly has 0 items instead of 1
Running X session wrapper
Loading profile from /etc/profile
Loading xinit script /etc/X11/xinit/xinitrc.d/40-libcanberra-gtk-module.sh
Loading xinit script /etc/X11/xinit/xinitrc.d/50-systemd-user.sh
X session wrapper complete, running session startlxqt
[Notice] Started

Mon Jul 29 15:29:14 2024
 Connections: accepted: [::1]::34178
 SConnection: Client needs protocol version 3.8
 SConnection: Client requests security type VeNCrypt(19)
 SVeNCrypt:   Client requests security type TLSVnc (258)

Mon Jul 29 15:29:15 2024
 VNCSConnST:  Server default pixel format depth 24 (32bpp) little-endian rgb888
```

**Comment by graysky2** [comment]
Thanks for the reply.  I modified my service like this:
```
- ExecStart=/usr/bin/vncserver -geometry 2000x1200 -localhost -alwaysshared %i
+ ExecStart=/usr/bin/vncserver -rendernode "" -geometry 2000x1200 -localhost -alwaysshared %i
```

I am not seeing anything in `$HOME/.local/state/tigervnc/familyroom:1.log`
```
Using desktop session lxqt

New 'familyroom:1 (facade)' desktop is familyroom:1

Starting desktop session lxqt


Xvnc TigerVNC 1.14.0 - built Jul 26 2024 23:15:30
Copyright (C) 1999-2024 TigerVNC Team and many others (see README.rst)
See https://www.tigervnc.org for information on TigerVNC.
Underlying X server release 12101013


Mon Jul 29 15:35:35 2024
 vncext:      VNC extension running!
 vncext:      Listening for VNC connections on all interface(s), port 5901
 vncext:      created VNC server for screen 0
[mi] mieq: warning: overriding existing handler (nil) with 0xaaaae0d9b188 for event 2
[mi] mieq: warning: overriding existing handler (nil) with 0xaaaae0d9b188 for event 3
xinit: XFree86_VT property unexpectedly has 0 items instead of 1
Running X session wrapper
Loading profile from /etc/profile
Loading xinit script /etc/X11/xinit/xinitrc.d/40-libcanberra-gtk-module.sh
Loading xinit script /etc/X11/xinit/xinitrc.d/50-systemd-user.sh
X session wrapper complete, running session startlxqt
[Notice] Started
xinit: connection to X server lost

waiting for X server to shut down 
Mon Jul 29 15:36:17 2024
 ComparingUpdateTracker: 0 pixels in / 0 pixels out
 ComparingUpdateTracker: (1:nan ratio)
```

**Comment by graysky2** [comment]
@oddfellow -
>I use the package provided systemd service file to start vncserver (/usr/lib/systemd/system/[vncserver@.service](mailto:vncserver@.service))

I tried starting with the package-provided systemd service and got the same result.

**Comment by CendioOssman** [comment]
Please add `rendernode=foobar` to `/etc/tigervnc/vncserver-config-defaults` and restart the service.

**Comment by graysky2** [comment]
@CendioOssman - that fixes things.  No more errors upon launching chromium and no visual artifacts.

**Comment by CendioOssman** [comment]
Thanks. That confirms there is some issue with the GPU acceleration. My first guess would be that the driver is not compensating for the tiling properly.

Unfortunately, I'm not able to reproduce the issue here. I'm getting a crash instead:

https://gitlab.freedesktop.org/mesa/mesa/-/issues/11594

Testing Arch Linux seems to involve a lot of manual steps. Is there a ready image available somewhere?

Could you share the output of `inxi -GSC -xx` to compare?

**Comment by CendioOssman** [comment]
Okay, I'm getting the same screen corruption when switching to a 64-bit Raspberry Pi OS. So it seems to be a 32-bit vs 64-bit issue.

Both bugs seem to have to do with tiling, though, so let's see what the mesa developers have to say.

**Comment by graysky2** [comment]
Thanks for your engagement with this and with mesa.

> Testing Arch Linux seems to involve a lot of manual steps. Is there a ready image available somewhere?

The official method installing Arch ARM is via zip file with some steps as you found.  To my knowledge there are not any official images.

> Could you share the output of inxi -GSC -xx to compare?

Yes:
```
% inxi -GSC -xx
System:
  Host: familyroom Kernel: 6.6.42-1-rpi arch: aarch64 bits: 64 compiler: gcc v: 14.1.1
  Console: pty pts/0 Distro: Arch Linux ARM
CPU:
  Info: quad core model: ARMv8 v8l variant: cortex-a72 bits: 64 type: MCP arch: v8l rev: 3 cache:
    L1: 320 KiB L2: 1024 KiB
  Speed (MHz): avg: 600 min/max: 600/1500 cores: 1: 600 2: 600 3: 600 4: 600 bogomips: 432
  Features: Use -f option to see features
Graphics:
  Device-1: bcm2711-hdmi0 driver: vc4_hdmi v: N/A bus-ID: N/A chip-ID: brcm:fef00700
  Device-2: bcm2711-hdmi1 driver: vc4_hdmi v: N/A bus-ID: N/A chip-ID: brcm:fef05700
  Device-3: bcm2711-vc5 driver: vc4_drm v: N/A bus-ID: N/A chip-ID: brcm:gpu
  Display: server: X.org v: 1.21.1.13 driver: X: loaded: modesetting alternate: fbdev dri: vc4
    gpu: vc4-drm,vc4_crtc,vc4_dpi,vc4_dsi,vc4_firmware_kms,vc4_hdmi,vc4_hvs,vc4_txp,vc4_v3d,vc4_vec
    tty: 128x36
  Monitor-1: HDMI-A-1 model: LG (GoldStar) TV res: 1920x1080 dpi: 30 diag: 1836mm (72.3")
  API: OpenGL Message: GL data unavailable in console, glxinfo missing.
```

**Comment by escuta** [comment]
> Please add `rendernode=foobar` to `/etc/tigervnc/vncserver-config-defaults` and restart the service.

Just to pipe in - this suggestion solves the problem for me on a Pi 4 with Arch linux (aarch64).
Thanks

**Comment by arthur-guru** [comment]
I get a similar rendering problem with latest Fedora 40/Gnome on a physical x86_64 machine using an oldish Nvidia Quadro 2000 card accessed via xrdp that uses the xvnc back-end. Problem doesn't exist when the machine is accessed directly.

Running `$ sudo dnf downgrade tigervnc-server-minimal tigervnc-license` downgrades from version 1.14.0-2 to 1.13.1-12 and fixes the problem.

The diagram below shows a screen grab of an xrdp session and in it is a Gnome Terminal with its screen rendered corruptly, however the Gnome background renders correctly. I have also right clicked the Gnome desktop to bring up a sub menu which renders correctly too.

![image](https://github.com/user-attachments/assets/3fe87aea-6ebc-43df-943e-a7eaa8a4bab0)

Happy to continue using the old version while waiting for a proper fix.

Thank you.

**Comment by CendioOssman** [comment]
The issue here is specific to the GPU in Raspberry Pi. So you'll need to report that issue to either Nvidia or nouveau, depending on what you are using.

**Comment by CendioOssman** [comment]
Upstream has now committed a fix. If anyone is up for it, you could try doing a custom build and see if it resolves everything.

**Comment by graysky2** [comment]
@CendioOssman - is the fix in a release?  If not, can you share the commit/s and I will try.

**Comment by graysky2** [comment]
I patched the tagged release 24.2.1 with that commit which I had to slightly rebase to apply and corresponding mesa package solved the issue.  I took care to remove the work-around line from `/etc/tigervnc/vncserver-config-defaults`

Thank you for reporting this, @CendioOssman.  I trust that commit will be in the next mesa release.  I think it's safe to close this.

**Comment by CendioOssman** [comment]
Great! Thanks for testing!
